src/journal/compress.c

```diff
@@ -541,6 +645,90 @@ int compress_stream_lz4(int fdf, int fdt, uint64_t max_bytes) {
 #endif
 }
 
+int compress_stream_zstd(int fdf, int fdt, uint64_t max_bytes) {
+#if HAVE_ZSTD
+        _cleanup_(ZSTD_freeCCtxp) ZSTD_CCtx *s = ZSTD_createCStream();
+        size_t out_size = ZSTD_CStreamOutSize();
+        uint8_t out[out_size];
+        char *src = NULL;
+        struct stat st;
+        ZSTD_inBuffer ZSTD_in = {src, 0, 0};
+        ZSTD_outBuffer ZSTD_out = {&out, sizeof(out), 0};
+        size_t ret;
+        uint64_t left = max_bytes;
+        int r;
+
+        assert(fdf >= 0);
+        assert(fdt >= 0);
+
+        if (!s)
+                return -ENOMEM;
+
+        ret = ZSTD_initCStream(s, ZSTD_COMPRESSION_LEVEL);
+        if (ZSTD_isError(ret)) {
+                log_error("Failed to initialize ZSTD encoder: %s", ZSTD_getErrorName(ret));
+                return zstd_ret_to_errnoh(ret);
+        }
+
+        if (fstat(fdf, &st) < 0)
+                return log_debug_errno(errno, "fstat() failed: %m");
+        ZSTD_in.size = st.st_size;
+
+        src = mmap(NULL, st.st_size, PROT_READ, MAP_PRIVATE, fdf, 0);
+        if (src == MAP_FAILED)
+                return -errno;
+        ZSTD_in.src = src;
+
+        for (;;) {
+                while (ZSTD_out.pos < ZSTD_out.size) {
+                        ret = ZSTD_compressStream(s, &ZSTD_out, &ZSTD_in);
+                        if (ZSTD_isError(ret)) {
+                                log_error("ZSTD encoder failed: %s", ZSTD_getErrorName(ret));
```

Comment:
in systemd, we try to avoid logging above LOG_DEBUG level in "library code", and this pretty much is "library code" (as opposed to "main program code"). See CODING_STYLE about this. Hence, please downgrade this to log_debug.